### DISCUSSION TITLE
`as_dict` performance problem in routing.py

### DISCUSSION THREAD

**Discussion by kigawas** [fastapi/fastapi#5669] [Questions] [answered]
## Description

In `routing.py`, there is an if-statement to check where the response is a `dataclass` instance: 

```py
def _prepare_response_content(
    res: Any,
    *,
    exclude_unset: bool,
    exclude_defaults: bool = False,
    exclude_none: bool = False,
) -> Any:
    if isinstance(res, BaseModel):
        ...
    elif isinstance(res, list):
        ...
    elif isinstance(res, dict):
        ...
    elif dataclasses.is_dataclass(res):
        return dataclasses.asdict(res)
    return res
```

The `dataclasses.asdict` is [notoriously slow](https://stackoverflow.com/questions/52229521/why-is-dataclasses-asdictobj-10x-slower-than-obj-dict), and for most of the cases, there is no need to convert to `dict` prematurely as it adds unnecessary overheads.

I'd rather just do something like:

```py
    if ...:
        ...
    elif dataclasses.is_dataclass(res):
        return res.__dict__
    return res
```

## Benchmark

```py
In [1]: import orjson

In [2]: from dataclasses import dataclass

In [3]: @dataclass
   ...: class A:
   ...:     a: int
   ...:     b: str
   ...:

In [4]: a = A(a=1,b='2')

In [5]: from dataclasses import asdict

In [6]: %timeit orjson.dumps(a)
203 ns ± 2.27 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [7]: %timeit orjson.dumps(asdict(a))
3.11 µs ± 36.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

As above, by removing `asdict` we can make it 10x faster in `_prepare_response_content`.

**Comment by kigawas** [comment]
Similarly, this also applies to the `jsonable_encoder` function in [`encoders.py`](https://github.com/tiangolo/fastapi/search?q=dataclasses.asdict):

```py
def jsonable_encoder(...):
    ...
    if dataclasses.is_dataclass(obj):
        obj_dict = dataclasses.asdict(obj)
    ...
```

**Comment by yinziyan1206** [comment]
if you do such trick, you cannot use some special type because of `JSONEncodeError`.

**Reply by kigawas** [reply]
Actually I don't need asdict since I'm using `orjson`.

**Comment by kigawas** [comment]
@tiangolo Do you have time to take a look on this? Or is there some "performance tuning guy" in the FastAPI team?

**Reply by kigawas** [reply]
I've proposed a PR: https://github.com/tiangolo/fastapi/pull/5743

**Comment by tiangolo** [comment, marked as answer]
This would be potentially dangerous as it might not account for some more complex use cases. If you want to improve performance, it would probably be better to use Pydantic.

In a future version FastAPI will use Pydantic's own serialization, equivalent to orjson, so that will improve a lot of things too. But thanks for the interest! ☕
